File: archipelago/src/archipelago.py
#!/usr/bin/python3

import sys

class DisjointSetForrest:
    def __init__(self, num_islands):
        self.num_islands = num_islands
        self.parents = [i for i in range(num_islands)]
        self.ranks = [0 for i in range(num_islands)]

    def find(self, elem):
        start_idx = elem
        idx = elem
        while self.parents[idx] != idx:
            idx = self.parents[idx]

        final_idx = idx
        idx = start_idx
        while self.parents[idx] != idx:
            parent_idx = self.parents[idx]
            self.parents[idx] = final_idx
            idx = parent_idx

        return final_idx

    def merge(self, elem_a, elem_b):
        head_a = self.find(elem_a)
        head_b = self.find(elem_b)

        if head_a == head_b:
            return

        if self.ranks[head_a] < self.ranks[head_b]:
            head_a, head_b = head_b, head_a

        self.parents[head_b] = head_a

        if self.ranks[head_a] == self.ranks[head_b]:
            self.ranks[head_a] += 1

def main():
    # Read in input
    N = int(input()) # Number of lines of input
    m = int(input()) # Number of islands

    dsforrest = DisjointSetForrest(m)

    for line in sys.stdin:
        fields = line.split()
        line_type = fields[0]
        first_island = int(fields[1])
        second_island = int(fields[2])

        if line_type == "B":
            print("Building {} {}".format(first_island, second_island))
            dsforrest.merge(first_island, second_island)
        elif line_type == "Q":
            print("Querying {} {}".format(first_island, second_island))
            if dsforrest.find(first_island) == dsforrest.find(second_island):
                print("Result: y")
            else:
                print("Result: n")
        else:
            print("Error: Unknown command type")

if __name__ == "__main__":
    main()

File: archipelago/src/archipelago.cpp

// Standard library includes
#include <iostream>
#include <vector>
#include <limits>
#include <sstream>

// Using declarations
using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::vector;

// Code for implementing disjoint sets

class DisjointSets {
    public:
        DisjointSets(unsigned int size);

        // Returns a representative element from the set containing
        // elem.
        int find(int elem);

        // Modifies the data structure to merge the sets containing a
        // and b
        void merge(int a, int b);

    private:
        std::vector<int> parents;
        std::vector<int> ranks;
};

typedef class DisjointSets DisjointSets;

DisjointSets::DisjointSets(unsigned int size) {
    for (int i = 0; i < size; i++) {
        this->parents.push_back(i);
        this->ranks.push_back(0);
    }
}

int DisjointSets::find(int elem) {
    int idx = elem;
    int start_idx = elem;
    int final_idx;

    while (idx != this->parents.at(idx)) {
        idx = this->parents.at(idx);
    }
    final_idx = idx;

    // Now run path compression
    idx = start_idx;
    while (idx != final_idx) {
        int next_idx;
        next_idx = this->parents.at(idx);
        this->parents.at(idx) = final_idx;
        idx = next_idx;
    }

    return final_idx;
}

void DisjointSets::merge(int a, int b) {
    int a_head = this->find(a);
    int b_head = this->find(b);

    if (a_head == b_head)
        return;

    if (this->ranks.at(a_head) < this->ranks.at(b_head)) {
        int tmp = a_head;
        a_head = b_head;
        b_head = tmp;
    }
    this->parents.at(b_head) = a_head;

    if (this->ranks.at(a_head) == this->ranks.at(b_head))
        this->ranks.at(a_head)++;
}

int main() {
    // Read in input
    int N; // number of lines in input
    int m; // number of islands
    cin >> N;
    cin >> m;
    cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    DisjointSets sets(m); // Initialize disjoint sets data structure

    for (string line; std::getline(cin, line); ) { // Loop over lines of input
        std::istringstream sline(line);
        char line_type;
        int arg1;
        int arg2;
        int q_result = -1;

        sline.get(line_type);
        sline >> arg1 >> arg2;
        if (line_type == 'B') {
            cout << "Building ";
            sets.merge(arg1, arg2);
        } else if (line_type == 'Q') {
            cout << "Querying ";
            if (sets.find(arg1) == sets.find(arg2))
                q_result = 1;
            else
                q_result = 0;
        } else {
            cout << "Unknown instruction ";
        }
        cout << arg1 << " " << arg2 << endl;
        if (q_result != -1)
            cout << "Result: " << ((q_result == 1) ? "y" : "n") << endl;
    }

    return 0;
}
